File: js/example_tts_long_input_compressed.js
#!/usr/bin/env node
/**
 * Example script for Inworld TTS synthesis with long text input (MP3 compressed).
 *
 * This script demonstrates how to synthesize speech from long text by:
 * 1. Chunking text at natural boundaries (paragraphs → newlines → sentences)
 * 2. Processing chunks through the TTS API with controlled concurrency
 * 3. Merging segment outputs with ffmpeg so duration/playback are correct
 */

const fs = require('fs');
const path = require('path');
const { execFileSync } = require('child_process');
const os = require('os');

// Configuration
const INPUT_FILE_PATH = '../tests-data/text/chapter1.txt';  // Path to input text file (relative to this script)
const MIN_CHUNK_SIZE = 500;   // Minimum characters before looking for break point
const MAX_CHUNK_SIZE = 1900;  // Maximum chunk size (API limit is 2000)
const MAX_CONCURRENT_REQUESTS = 2;  // Limit parallel requests to avoid RPS limits
const MAX_RETRIES = 3;        // Maximum retries for rate limit errors
const RETRY_BASE_DELAY = 1000; // Base delay for exponential backoff (ms)

// Audio configuration for MP3
const SAMPLE_RATE = 48000;

/**
 * Check if INWORLD_API_KEY environment variable is set.
 * @returns {string|null} API key or null if not set
 */
function checkApiKey() {
    const apiKey = process.env.INWORLD_API_KEY;
    if (!apiKey) {
        console.log('Error: INWORLD_API_KEY environment variable is not set.');
        console.log('Please set it with: export INWORLD_API_KEY=your_api_key_here');
        return null;
    }
    return apiKey;
}

/**
 * Find the best break point in text, prioritizing:
 * 1. Paragraph breaks (\n\n)
 * 2. Line breaks (\n)
 * 3. Sentence endings (.!?)
 * 4. Last space (fallback)
 * 
 * @param {string} text - Text to search for break point
 * @param {number} minPos - Minimum position to start looking
 * @param {number} maxPos - Maximum position to look
 * @param {number} chunkIndex - Current chunk index for logging
 * @returns {number} Position of best break point
 */
function findBreakPoint(text, minPos, maxPos, chunkIndex) {
    const searchText = text.slice(0, maxPos);
    
    // 1. Try paragraph breaks (\n\n) after minPos
    let breakIndex = -1;
    let searchStart = minPos;
    
    while (true) {
        const idx = searchText.indexOf('\n\n', searchStart);
        if (idx === -1 || idx >= maxPos) break;
        if (idx >= minPos) {
            breakIndex = idx + 2; // Include the paragraph break
            break; // Use first paragraph break after minPos
        }
        searchStart = idx + 1;
    }
    
    if (breakIndex > 0) {
        console.log(`  Chunk ${chunkIndex + 1}: Found paragraph break at position ${breakIndex}`);
        return breakIndex;
    }
    
    // 2. Try single line breaks (\n) after minPos
    searchStart = minPos;
    while (true) {
        const idx = searchText.indexOf('\n', searchStart);
        if (idx === -1 || idx >= maxPos) break;
        if (idx >= minPos) {
            breakIndex = idx + 1;
            break;
        }
        searchStart = idx + 1;
    }
    
    if (breakIndex > 0) {
        console.log(`  Chunk ${chunkIndex + 1}: Found line break at position ${breakIndex}`);
        return breakIndex;
    }
    
    // 3. Try sentence endings after minPos
    const sentenceEndRegex = /[.!?]["'"']?\s+|[.!?]["'"']?$/g;
    let match;
    while ((match = sentenceEndRegex.exec(searchText)) !== null) {
        if (match.index >= minPos) {
            console.log(`  Chunk ${chunkIndex + 1}: Found sentence break at position ${match.index + match[0].length}`);
            return match.index + match[0].length;
        }
    }
    
    // 4. Fall back to any sentence end before maxPos
    sentenceEndRegex.lastIndex = 0;
    let lastSentenceEnd = -1;
    while ((match = sentenceEndRegex.exec(searchText)) !== null) {
        lastSentenceEnd = match.index + match[0].length;
    }
    if (lastSentenceEnd > 0) {
        console.log(`  Chunk ${chunkIndex + 1}: Found sentence break (fallback) at position ${lastSentenceEnd}`);
        return lastSentenceEnd;
    }
    
    // 5. Last resort: break at last space
    const spaceIndex = searchText.lastIndexOf(' ');
    const breakPos = spaceIndex > 0 ? spaceIndex + 1 : maxPos;
    console.log(`  Chunk ${chunkIndex + 1}: Found space break (fallback) at position ${breakPos}`);
    return breakPos;
}

/**
 * Chunk text into segments at natural boundaries.
 * Prioritizes paragraph breaks, then line breaks, then sentence endings.
 * 
 * @param {string} text - The full text to chunk
 * @returns {Array<{text: string, startChar: number, endChar: number}>} Array of text chunks
 */
function chunkText(text) {
    const chunks = [];
    let currentPosition = 0;
    
    while (currentPosition < text.length) {
        const remainingText = text.slice(currentPosition);
        
        // If remaining text is short enough, take it all
        if (remainingText.length <= MAX_CHUNK_SIZE) {
            const chunkContent = remainingText.trim();
            if (chunkContent.length > 0) {
                chunks.push({
                    text: chunkContent,
                    startChar: currentPosition,
                    endChar: text.length
                });
            }
            break;
        }
        
        // Find the best break point
        const chunkEnd = findBreakPoint(remainingText, MIN_CHUNK_SIZE, MAX_CHUNK_SIZE, chunks.length);
        
        const chunkContent = remainingText.slice(0, chunkEnd).trim();
        if (chunkContent.length > 0) {
            chunks.push({
                text: chunkContent,
                startChar: currentPosition,
                endChar: currentPosition + chunkEnd
            });
        }
        
        currentPosition += chunkEnd;
    }
    
    return chunks;
}

/**
 * Sleep for a specified duration.
 * @param {number} ms - Milliseconds to sleep
 * @returns {Promise<void>}
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Synthesize speech from text using Inworld TTS API with retry logic.
 * 
 * @param {string} text - Text to synthesize
 * @param {string} voiceId - Voice ID to use
 * @param {string} modelId - Model ID to use
 * @param {string} apiKey - API key for authentication
 * @param {number} chunkIndex - Chunk index for logging
 * @param {number} totalChunks - Total number of chunks
 * @returns {Promise<Buffer>} Audio data
 */
async function synthesizeSpeech(text, voiceId, modelId, apiKey, chunkIndex, totalChunks) {
    const url = 'https://api.inworld.ai/tts/v1/voice';
    
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Basic ${apiKey}`
    };
    
    const requestData = {
        text: text,
        voice_id: voiceId,
        model_id: modelId,
        audio_config: {
            audio_encoding: 'MP3',
            sample_rate_hertz: SAMPLE_RATE
        }
    };
    
    for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
        try {
            console.log(`[${chunkIndex + 1}/${totalChunks}] Synthesizing chunk (${text.length} chars)...`);
            
            const response = await fetch(url, {
                method: 'POST',
                headers: headers,
                body: JSON.stringify(requestData)
            });
            
            // Check for rate limit or other HTTP errors
            if (!response.ok) {
                const isRateLimit = response.status === 429;
                const isLastAttempt = attempt === MAX_RETRIES - 1;
                
                if (isRateLimit && !isLastAttempt) {
                    const delay = RETRY_BASE_DELAY * Math.pow(2, attempt);
                    console.log(`[${chunkIndex + 1}/${totalChunks}] Rate limited, retrying in ${delay}ms...`);
                    await sleep(delay);
                    continue;
                }
                
                let errorDetails = '';
                try {
                    const errorData = await response.json();
                    errorDetails = JSON.stringify(errorData);
                } catch {
                    errorDetails = await response.text();
                }
                throw new Error(`HTTP ${response.status}: ${errorDetails}`);
            }
            
            const result = await response.json();
            const audioData = Buffer.from(result.audioContent, 'base64');
            
            console.log(`[${chunkIndex + 1}/${totalChunks}] Done - ${audioData.length} bytes`);
            
            return audioData;
            
        } catch (error) {
            const isRateLimit = error.message.includes('429');
            const isLastAttempt = attempt === MAX_RETRIES - 1;
            
            if (isRateLimit && !isLastAttempt) {
                const delay = RETRY_BASE_DELAY * Math.pow(2, attempt);
                console.log(`[${chunkIndex + 1}/${totalChunks}] Rate limited, retrying in ${delay}ms...`);
                await sleep(delay);
                continue;
            }
            
            console.log(`Error for chunk ${chunkIndex + 1}: ${error.message}`);
            if (error.response) {
                try {
                    console.log(`   Error details: ${JSON.stringify(error.response.data)}`);
                } catch {
                    console.log(`   Response text: ${error.response.data}`);
                }
            }
            throw error;
        }
    }
}

/**
 * Process chunks with controlled concurrency.
 * @param {Array} chunks - Text chunks to process
 * @param {string} voiceId - Voice ID
 * @param {string} modelId - Model ID
 * @param {string} apiKey - API key
 * @returns {Promise<Array<Buffer>>} Audio buffers in order
 */
async function synthesizeAllChunks(chunks, voiceId, modelId, apiKey) {
    const results = new Array(chunks.length);
    const queue = chunks.map((chunk, index) => ({ chunk, index }));
    
    async function processNext() {
        while (queue.length > 0) {
            const { chunk, index } = queue.shift();
            results[index] = await synthesizeSpeech(
                chunk.text,
                voiceId,
                modelId,
                apiKey,
                index,
                chunks.length
            );
        }
    }
    
    // Start concurrent workers
    const workers = [];
    for (let i = 0; i < Math.min(MAX_CONCURRENT_REQUESTS, chunks.length); i++) {
        workers.push(processNext());
    }
    
    await Promise.all(workers);
    return results;
}

/**
 * Combine multiple audio buffers into one by raw concatenation.
 * Note: Each API response is a complete MP3 file (with its own header/duration). Raw concat
 * produces a file that players and duration tools interpret as only the first segment (e.g. 1:00),
 * so the reported duration/playback metadata may be incorrect.
 *
 * Prefer mergeMp3SegmentsWithFfmpeg() when ffmpeg is available, as it produces a single MP3 with
 * correct duration and playback. This function is intended as a fallback when ffmpeg is not
 * available or cannot be used.
 * @param {Array<Buffer>} audioBuffers - Audio buffers to combine
 * @returns {Buffer} Combined audio
 */
function combineAudioBuffers(audioBuffers) {
    return Buffer.concat(audioBuffers);
}

/**
 * Merge multiple MP3 buffers into one file with correct duration using ffmpeg.
 * Each non-streaming API response is a full MP3; raw concat makes duration/show length wrong.
 * @param {Array<Buffer>} audioBuffers - One MP3 buffer per segment
 * @param {string} outputFile - Output path for merged MP3
 * @returns {boolean} true if merged with ffmpeg, false if ffmpeg unavailable (caller should fall back)
 */
function mergeMp3SegmentsWithFfmpeg(audioBuffers, outputFile) {
    const tmpDir = path.join(os.tmpdir(), `inworld-tts-long-${Date.now()}`);
    fs.mkdirSync(tmpDir, { recursive: true });
    try {
        const listPath = path.join(tmpDir, 'list.txt');
        const segPaths = [];
        for (let i = 0; i < audioBuffers.length; i++) {
            const segPath = path.join(tmpDir, `seg_${i}.mp3`);
            fs.writeFileSync(segPath, audioBuffers[i]);
            segPaths.push(segPath);
        }
        // ffmpeg concat demuxer: paths must be escaped for ' in file names
        const listContent = segPaths.map((p) => `file '${p.replace(/'/g, "'\\''")}'`).join('\n');
        fs.writeFileSync(listPath, listContent);
        execFileSync('ffmpeg', ['-y', '-f', 'concat', '-safe', '0', '-i', listPath, '-c', 'copy', outputFile], {
            stdio: ['ignore', 'pipe', 'pipe']
        });
        return true;
    } catch (e) {
        const ffmpegNotFound = e && (e.code === 'ENOENT' || (e.message && e.message.includes('ffmpeg')));
        if (ffmpegNotFound) {
            console.log('   (ffmpeg not found; saving raw concatenation — duration may show as first segment only)');
        } else {
            console.error('   (ffmpeg failed while merging MP3 segments; falling back to raw concatenation)');
            console.error('   Error:', e && e.message ? e.message : e);
            if (e && e.stderr) {
                try {
                    const stderrStr = typeof e.stderr === 'string' ? e.stderr : e.stderr.toString();
                    if (stderrStr) {
                        console.error('   ffmpeg stderr:\n' + stderrStr);
                    }
                } catch (_) {}
            }
        }
        return false;
    } finally {
        try {
            const files = fs.readdirSync(tmpDir);
            for (const f of files) fs.unlinkSync(path.join(tmpDir, f));
            fs.rmdirSync(tmpDir);
        } catch (_) {}
    }
}

/**
 * Save audio to file.
 * @param {Buffer} audioData - Audio data (MP3)
 * @param {string} outputFile - Output file path
 */
function saveAudioToFile(audioData, outputFile) {
    fs.writeFileSync(outputFile, audioData);
    console.log(`Audio saved to: ${outputFile}`);
}

/**
 * Read input text from file.
 * @param {string} inputFile - Path to input file
 * @returns {string|null} Text content or null on error
 */
function readInputText(inputFile) {
    try {
        const text = fs.readFileSync(inputFile, 'utf-8');
        console.log(`Loaded: ${inputFile} (${text.length} chars)`);
        return text;
    } catch (error) {
        console.log(`Error reading input file: ${error.message}`);
        return null;
    }
}

/**
 * Main function - high-level orchestration only.
 */
async function main() {
    console.log('Inworld TTS Long Text Synthesis (MP3 Compressed)\n');
    
    // Setup
    const apiKey = checkApiKey();
    if (!apiKey) return 1;
    
    // Configuration - modify these for your use case
    const voiceId = 'Edward';
    const modelId = 'inworld-tts-1.5-max';
    const outputFile = 'synthesis_long_output.mp3';
    const inputFile = path.join(__dirname, INPUT_FILE_PATH);
    
    // Read input text
    const text = readInputText(inputFile);
    if (!text) return 1;
    
    // Split into chunks
    const chunks = chunkText(text);
    console.log(`Split into ${chunks.length} chunks (min: ${MIN_CHUNK_SIZE}, max: ${MAX_CHUNK_SIZE} chars)\n`);
    
    try {
        const startTime = Date.now();
        
        // Synthesize all chunks
        console.log(`Synthesizing with ${MAX_CONCURRENT_REQUESTS} concurrent requests...\n`);
        const audioBuffers = await synthesizeAllChunks(chunks, voiceId, modelId, apiKey);
        
        // Merge audio (ffmpeg gives correct duration; raw concat would show first segment only)
        console.log('\nMerging audio...');
        const mergedWithFfmpeg = mergeMp3SegmentsWithFfmpeg(audioBuffers, outputFile);
        if (!mergedWithFfmpeg) {
            const combinedAudio = combineAudioBuffers(audioBuffers);
            saveAudioToFile(combinedAudio, outputFile);
        } else {
            console.log(`Audio saved to: ${outputFile}`);
        }

        const fileSizeKB = (fs.statSync(outputFile).size / 1024).toFixed(1);
        const elapsed = (Date.now() - startTime) / 1000;
        console.log(`Output size: ${fileSizeKB} KB`);
        console.log(`Completed in ${elapsed.toFixed(2)}s`);
        
    } catch (error) {
        console.log(`\nSynthesis failed: ${error.message}`);
        return 1;
    }
    
    return 0;
}

if (require.main === module) {
    main().then(process.exit);
}

module.exports = { chunkText, synthesizeSpeech, combineAudioBuffers, mergeMp3SegmentsWithFfmpeg, synthesizeAllChunks };

File: tts/python/example_tts_long_input_compressed.py
#!/usr/bin/env python3
"""
Example script for Inworld TTS synthesis with long text input (MP3 compressed).

This script demonstrates how to synthesize speech from long text by:
1. Chunking text at natural boundaries (paragraphs → newlines → sentences)
2. Processing chunks through the TTS API with controlled concurrency
3. Merging segment outputs with ffmpeg so duration/playback are correct.
"""

import base64
import os
import re
import subprocess
import tempfile
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass  # dotenv is optional; INWORLD_API_KEY can also be set via export

import requests

# Configuration
INPUT_FILE_PATH = "../tests-data/text/chapter1.txt"  # Path to input text file (relative to this script)
MIN_CHUNK_SIZE = 500    # Minimum characters before looking for break point
MAX_CHUNK_SIZE = 1600   # Maximum chunk size (API limit is 2000)
MAX_CONCURRENT_REQUESTS = 2   # Limit parallel requests to avoid RPS limits
MAX_RETRIES = 3         # Maximum retries for rate limit errors
RETRY_BASE_DELAY = 1.0  # Base delay for exponential backoff (seconds)
CHARS_PER_SECOND = 12.0 # Approx speaking rate; used to convert <break> durations to equivalent char counts
CJK_CHAR_WEIGHT = 3.0   # CJK chars produce ~3x more audio than Latin chars

# Audio configuration for MP3
SAMPLE_RATE = 48000


@dataclass
class TextChunk:
    """Represents a chunk of text with its position in the original text."""
    text: str
    start_char: int
    end_char: int


@dataclass
class SynthesisConfig:
    """Configuration for TTS synthesis."""
    voice_id: str
    model_id: str
    api_key: str
    audio_encoding: str = "MP3"


def check_api_key() -> Optional[str]:
    """Check if INWORLD_API_KEY environment variable is set."""
    api_key = os.getenv("INWORLD_API_KEY")
    if not api_key:
        print("Error: INWORLD_API_KEY environment variable is not set.")
        print("Please set it with: export INWORLD_API_KEY=your_api_key_here")
        return None
    return api_key


def _count_cjk(text: str) -> int:
    """Count CJK characters (Chinese, Japanese Kanji/Kana, Korean Hangul)."""
    return sum(1 for c in text if '\u4e00' <= c <= '\u9fff'
               or '\u3040' <= c <= '\u30ff' or '\uac00' <= c <= '\ud7af')


def estimate_effective_length(text: str, chars_per_second: float = CHARS_PER_SECOND) -> int:
    """Return text length with SSML <break> durations and CJK weight factored in.

    Without this, chunks stuffed with <break time="Xs"/> tags look short in chars
    but produce huge audio, potentially exceeding API response size limits. CJK characters also
    produce ~3x more audio per character than Latin text.
    """
    break_pattern = re.compile(r'<break\s+time="([\d.]+)(m?s)"\s*/?>', re.IGNORECASE)

    total_break_seconds = 0.0
    for m in break_pattern.finditer(text):
        try:
            val = float(m.group(1))
            total_break_seconds += val / 1000 if m.group(2).lower() == 'ms' else val
        except ValueError:
            pass

    text_without_breaks = re.sub(r'<break\s[^>]*/?>',  '', text, flags=re.IGNORECASE)
    cjk_count = _count_cjk(text_without_breaks)
    raw_length = (len(text_without_breaks) - cjk_count) + int(cjk_count * CJK_CHAR_WEIGHT)

    return raw_length + int(total_break_seconds * chars_per_second)


def find_break_point(text: str, min_pos: int, max_pos: int, chunk_index: int) -> int:
    """
    Find the best break point in text, prioritizing:
    1. Paragraph breaks (\\n\\n)
    2. Line breaks (\\n)
    3. Sentence endings (.!?)
    4. Last space (fallback)
    
    Args:
        text: Text to search for break point
        min_pos: Minimum position to start looking
        max_pos: Maximum position to look
        chunk_index: Current chunk index for logging
        
    Returns:
        Position of best break point
    """
    search_text = text[:max_pos]
    
    # 1. Try paragraph breaks (\n\n) after min_pos
    search_start = min_pos
    while True:
        idx = search_text.find('\n\n', search_start)
        if idx == -1 or idx >= max_pos:
            break
        if idx >= min_pos:
            print(f"  Chunk {chunk_index + 1}: Found paragraph break at position {idx + 2}")
            return idx + 2  # Include the paragraph break
        search_start = idx + 1
    
    # 2. Try single line breaks (\n) after min_pos
    search_start = min_pos
    while True:
        idx = search_text.find('\n', search_start)
        if idx == -1 or idx >= max_pos:
            break
        if idx >= min_pos:
            print(f"  Chunk {chunk_index + 1}: Found line break at position {idx + 1}")
            return idx + 1
        search_start = idx + 1
    
    # 3. Try sentence endings after min_pos
    sentence_end_pattern = re.compile(r'[.!?]["\'"\'"]?\s+|[.!?]["\'"\'"]?$')
    for match in sentence_end_pattern.finditer(search_text):
        if match.start() >= min_pos:
            print(f"  Chunk {chunk_index + 1}: Found sentence break at position {match.end()}")
            return match.end()
    
    # 4. Fall back to any sentence end before max_pos
    last_sentence_end = -1
    for match in sentence_end_pattern.finditer(search_text):
        last_sentence_end = match.end()
    if last_sentence_end > 0:
        print(f"  Chunk {chunk_index + 1}: Found sentence break (fallback) at position {last_sentence_end}")
        return last_sentence_end
    
    # 5. Last resort: break at last space
    space_index = search_text.rfind(' ')
    break_pos = space_index + 1 if space_index > 0 else max_pos
    print(f"  Chunk {chunk_index + 1}: Found space break (fallback) at position {break_pos}")
    return break_pos


def chunk_text(text: str) -> list[TextChunk]:
    """
    Chunk text into segments at natural boundaries.
    Prioritizes paragraph breaks, then line breaks, then sentence endings.
    
    Args:
        text: The full text to chunk
        
    Returns:
        List of TextChunk objects with text and positions
    """
    chunks = []
    current_position = 0
    
    while current_position < len(text):
        remaining_text = text[current_position:]
        
        # If remaining text fits within the effective budget, take it all
        if estimate_effective_length(remaining_text) <= MAX_CHUNK_SIZE:
            chunk_content = remaining_text.strip()
            if chunk_content:
                chunks.append(TextChunk(
                    text=chunk_content,
                    start_char=current_position,
                    end_char=len(text)
                ))
            break
        
        # Shrink search window when <break> tags inflate effective length
        candidate = remaining_text[:MAX_CHUNK_SIZE]
        eff_len = estimate_effective_length(candidate)
        if eff_len > MAX_CHUNK_SIZE:
            scale = MAX_CHUNK_SIZE / eff_len
            effective_max = max(1, int(len(candidate) * scale))
            effective_min = max(1, int(MIN_CHUNK_SIZE * scale))
        else:
            effective_max = MAX_CHUNK_SIZE
            effective_min = MIN_CHUNK_SIZE

        chunk_end = find_break_point(remaining_text, effective_min, effective_max, len(chunks))
        
        chunk_content = remaining_text[:chunk_end].strip()
        if chunk_content:
            chunks.append(TextChunk(
                text=chunk_content,
                start_char=current_position,
                end_char=current_position + chunk_end
            ))
        
        current_position += chunk_end
    
    return chunks


def synthesize_speech(
    text: str,
    config: SynthesisConfig,
    chunk_index: int,
    total_chunks: int
) -> bytes:
    """
    Synthesize speech from text using Inworld TTS API with retry logic.
    
    Args:
        text: Text to synthesize
        config: Synthesis configuration
        chunk_index: Index of this chunk (for logging)
        total_chunks: Total number of chunks (for logging)
        
    Returns:
        bytes: Audio data
    """
    url = "https://api.inworld.ai/tts/v1/voice"
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Basic {config.api_key}"
    }
    
    request_data = {
        "text": text,
        "voice_id": config.voice_id,
        "model_id": config.model_id,
        "audio_config": {
            "audio_encoding": config.audio_encoding,
            "sample_rate_hertz": SAMPLE_RATE
        }
    }
    
    for attempt in range(MAX_RETRIES):
        try:
            print(f"[{chunk_index + 1}/{total_chunks}] Synthesizing chunk ({len(text)} chars)...")
            
            response = requests.post(url, headers=headers, json=request_data)
            response.raise_for_status()
            
            result = response.json()
            audio_data = base64.b64decode(result["audioContent"])
            
            print(f"[{chunk_index + 1}/{total_chunks}] Done - {len(audio_data)} bytes")
            return audio_data
            
        except requests.exceptions.RequestException as e:
            is_rate_limit = hasattr(e, 'response') and e.response is not None and e.response.status_code == 429
            is_last_attempt = attempt == MAX_RETRIES - 1
            
            if is_rate_limit and not is_last_attempt:
                delay = RETRY_BASE_DELAY * (2 ** attempt)
                print(f"[{chunk_index + 1}/{total_chunks}] Rate limited, retrying in {delay:.1f}s...")
                time.sleep(delay)
                continue
            
            print(f"Error for chunk {chunk_index + 1}: {e}")
            if hasattr(e, 'response') and e.response is not None:
                try:
                    error_detail = e.response.json()
                    print(f"   Error details: {error_detail}")
                except Exception:
                    print(f"   Response text: {e.response.text}")
            raise
    
    # Should not reach here, but just in case
    raise RuntimeError(f"Failed to synthesize chunk {chunk_index + 1} after {MAX_RETRIES} attempts")


def synthesize_chunk_wrapper(args: tuple) -> tuple[int, bytes]:
    """Wrapper function for parallel synthesis."""
    chunk, config, index, total = args
    return index, synthesize_speech(chunk.text, config, index, total)


def synthesize_all_chunks(
    chunks: list[TextChunk],
    config: SynthesisConfig
) -> list[bytes]:
    """
    Process chunks with controlled concurrency.
    
    Args:
        chunks: Text chunks to process
        config: Synthesis configuration
        
    Returns:
        Audio buffers in order
    """
    synthesis_args = [
        (chunk, config, i, len(chunks))
        for i, chunk in enumerate(chunks)
    ]
    
    audio_buffers = [None] * len(chunks)
    
    with ThreadPoolExecutor(max_workers=MAX_CONCURRENT_REQUESTS) as executor:
        futures = {
            executor.submit(synthesize_chunk_wrapper, args): args[2]
            for args in synthesis_args
        }
        
        for future in as_completed(futures):
            index, audio_data = future.result()
            audio_buffers[index] = audio_data
    
    return audio_buffers


def combine_audio_buffers(audio_buffers: list[bytes]) -> bytes:
    """
    Combine multiple MP3 audio buffers into one by raw concatenation.
    Note: Each API response is a complete MP3 file (with its own header/duration). Raw concat
    produces a file that players and duration tools interpret as only the first segment,
    so the reported duration/playback metadata may be incorrect.

    Prefer merge_mp3_segments_with_ffmpeg() when ffmpeg is available, as it produces a single
    MP3 with correct duration and playback. This function is intended as a fallback when ffmpeg
    is not available or cannot be used.
    """
    return b''.join(audio_buffers)


def merge_mp3_segments_with_ffmpeg(audio_buffers: list[bytes], output_file: str) -> bool:
    """
    Merge multiple MP3 buffers into one file with correct duration using ffmpeg.
    Each non-streaming API response is a full MP3; raw concat makes duration/show length wrong.
    Returns True if merged with ffmpeg, False if ffmpeg unavailable (caller should fall back).
    """
    tmp_dir = tempfile.mkdtemp(prefix="inworld-tts-long-")
    try:
        list_path = os.path.join(tmp_dir, "list.txt")
        seg_paths = []
        for i, buf in enumerate(audio_buffers):
            seg_path = os.path.join(tmp_dir, f"seg_{i}.mp3")
            with open(seg_path, "wb") as f:
                f.write(buf)
            seg_paths.append(seg_path)
        with open(list_path, "w") as f:
            for p in seg_paths:
                escaped_p = p.replace("'", "'\\''")
                f.write(f"file '{escaped_p}'\n")
        result = subprocess.run(
            ["ffmpeg", "-y", "-f", "concat", "-safe", "0", "-i", list_path, "-c", "copy", output_file],
            capture_output=True,
            timeout=120,
        )
        if result.returncode == 0:
            return True
        return False
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return False
    except Exception:
        return False
    finally:
        try:
            for f in os.listdir(tmp_dir):
                os.unlink(os.path.join(tmp_dir, f))
            os.rmdir(tmp_dir)
        except OSError:
            pass


def save_audio_to_file(audio_data: bytes, output_file: str):
    """Save MP3 audio to file."""
    with open(output_file, 'wb') as f:
        f.write(audio_data)
    print(f"Audio saved to: {output_file}")


def read_input_text(input_file: Path) -> Optional[str]:
    """Read input text from file."""
    try:
        text = input_file.read_text(encoding="utf-8")
        print(f"Loaded: {input_file} ({len(text)} chars)")
        return text
    except Exception as e:
        print(f"Error reading input file: {e}")
        return None


def main():
    """Main function - high-level orchestration only."""
    print("Inworld TTS Long Text Synthesis (MP3 Compressed)\n")
    
    # Setup
    api_key = check_api_key()
    if not api_key:
        return 1
    
    # Configuration - modify these for your use case
    voice_id = "Edward"
    model_id = "inworld-tts-1.5-max"  # max model for non-streaming: higher quality
    output_file = "synthesis_long_output.mp3"
    
    script_dir = Path(__file__).parent
    input_file = script_dir / INPUT_FILE_PATH
    
    config = SynthesisConfig(
        voice_id=voice_id,
        model_id=model_id,
        api_key=api_key,
        audio_encoding="MP3"
    )
    
    # Read input text
    text = read_input_text(input_file)
    if not text:
        return 1
    
    # Split into chunks
    chunks = chunk_text(text)
    print(f"Split into {len(chunks)} chunks (min: {MIN_CHUNK_SIZE}, max: {MAX_CHUNK_SIZE} chars)\n")
    
    try:
        start_time = time.time()
        
        # Synthesize all chunks
        print(f"Synthesizing with {MAX_CONCURRENT_REQUESTS} concurrent requests...\n")
        audio_buffers = synthesize_all_chunks(chunks, config)
        
        # Merge audio (ffmpeg gives correct duration; raw concat would show first segment only)
        print("\nMerging audio...")
        merged_with_ffmpeg = merge_mp3_segments_with_ffmpeg(audio_buffers, output_file)
        if not merged_with_ffmpeg:
            print("   (ffmpeg unavailable or failed; saving raw concatenation — duration may show as first segment only)")
            combined_audio = combine_audio_buffers(audio_buffers)
            save_audio_to_file(combined_audio, output_file)
        else:
            print(f"Audio saved to: {output_file}")
        
        # Report
        file_size_kb = os.path.getsize(output_file) / 1024
        elapsed = time.time() - start_time
        print(f"Output size: {file_size_kb:.1f} KB")
        print(f"Completed in {elapsed:.2f}s")
        
    except Exception as e:
        print(f"\nSynthesis failed: {e}")
        return 1
    
    return 0


if __name__ == "__main__":
    exit(main())

